src/shares/PaymentHub.sol

```diff
@@ -175,10 +176,11 @@ contract PaymentHub {
     function payFromEtherAndNotify(address recipient, uint256 xchfamount, bytes calldata ref) payable external {
         // Check if the brokerbot has setting to keep ETH
         if (hasSettingKeepEther(recipient)) {
-            uint256 priceInEther = getPriceInEther(xchfamount, recipient);
+            uint256 priceInEther = getPriceInEtherFromOracle(xchfamount);
 
             // If ETH send in the transaction is smaller than current price revert
-            require(msg.value >= priceInEther, "not enough ether");
+            // Note Luzius: this require statement is unnecessary, as the subsequent call would fail anyway
```

Comment:
Are you sure? **processIncoming**() is called, that calls **buy**() just assumes **amount** number of **base** tokens (or their value in ETH) have already been transferred and sends out the shares. I may be wrong but pls check again.